Sources/Basics/Observability.swift

```diff
@@ -0,0 +1,293 @@
+        self.underlying = underlying
+    }
+
+    public var severity: DiagnosticMessage.Severity {
+        self.underlying.severity
+    }
+
+    public var message: String {
+        self.underlying.message
+    }
+
+
+    public var data: CustomStringConvertible? {
+        self.underlying.data
+    }
+
+    public var description: String {
+        return "[\(self.severity)]: \(self.message)"
+    }
+
+    public static func == (lhs: Diagnostic, rhs: Diagnostic) -> Bool {
+        if lhs.context?.description != rhs.context?.description {
+            return false
+        }
+        if lhs.underlying != rhs.underlying {
+            return false
+        }
+        return true
+    }
+}
+
+// TODO: consider using @autoclosure to delay potentially expensive evaluation of data when some diagnostics may be filtered out
+public struct DiagnosticsEmitter {
+    public let context: DiagnosticsContext?
+    private let handler: DiagnosticsHandler
+
+    public init(context: DiagnosticsContext? = .none) {
+        self.context = context
+        self.handler = ObservabilitySystem.global.diagnosticsHandler
+    }
+
+    public func emit(_ diagnostic: Diagnostic) {
+        self.handler(diagnostic)
+    }
+
+    public func emit(_ message: DiagnosticMessage) {
+        self.emit(.init(context: self.context, message: message))
+    }
+
+    public func emit(severity: DiagnosticMessage.Severity, message: String, data: CustomStringConvertible? = .none) {
+        self.emit(.init(context: self.context, message: .init(severity: severity, message: message, data: data)))
+    }
+
+    public func emit(error message: String, data: CustomStringConvertible? = .none) {
+        self.emit(.error(message, data: data))
+    }
+
+    public func emit(_ error: Error, data: CustomStringConvertible? = .none) {
+        // FIXME: this brings in the TSC API still
+        if self.context == nil, let errorProvidingLocation = error as? DiagnosticLocationProviding, let diagnosticLocation = errorProvidingLocation.diagnosticLocation {
+            let context = DiagnosticLocationWrapper(diagnosticLocation)
+            return DiagnosticsEmitter(context: context).emit(error)
+        }
+        self.emit(.error(error, data: data))
+    }
+
+    public func emit(warning message: String, data: CustomStringConvertible? = .none) {
+        self.emit(.warning(message, data: data))
+    }
+
+    public func emit(info message: String, data: CustomStringConvertible? = .none) {
+        self.emit(.info(message, data: data))
+    }
+
+    public func trap<T>(_ closure: () throws -> T) -> T? {
+        do  {
+            return try closure()
+        } catch {
+            self.emit(error)
+            return nil
+        }
+    }
+}
+
+public struct DiagnosticMessage: Equatable {
+    let severity: Severity
+    let message: String
+    // 👀 TODO: not sure if this is used very much, but it we need it this could/should be changed to more structured metadata model
```

Comment:
Absolutely agreed.  What would be the best, I think, is to have the diagnostic contain the underlying information and then have formatters that can present it in either console form or in an IDE.  I think we will still need textual information here in the form of a message, but maybe the conversion to a flat string is something that should be left to decorators.